key_cacher.go

```diff
@@ -3,19 +3,61 @@ package jose
 import (
 	b64 "encoding/base64"
 	"errors"
+	"fmt"
+	"sync"
 	"time"
 
+	"github.com/luraproject/lura/v2/config"
+	"github.com/luraproject/lura/v2/logging"
 	"gopkg.in/square/go-jose.v2"
 )
 
 var (
-	ErrNoKeyFound = errors.New("no Keys have been found")
-	ErrKeyExpired = errors.New("key exists but is expired")
+	ErrNoKeyFound            = errors.New("no Keys have been found")
+	ErrKeyExpired            = errors.New("key exists but is expired")
+	defaultGlobalCacheMaxAge = 15 * time.Minute
 
 	// Configuring with MaxKeyAgeNoCheck will skip key expiry check
-	MaxKeyAgeNoCheck = time.Duration(-1)
+	MaxKeyAgeNoCheck    = time.Duration(-1)
+	globalKeyCacher     = map[string]GlobalCacher{}
+	globalKeyCacherOnce = new(sync.Once)
 )
 
+type GlobalCacher struct {
+	kc KeyCacher
+	mu *sync.RWMutex
+}
+
+func SetGlobalCacher(l logging.Logger, cfg config.ExtraConfig) error {
+	duration, err := configGetter(cfg)
+	if err != nil {
+		if err != ErrNoValidatorCfg {
+			l.Error("[SERVICE: JWTValidator]", err.Error())
+		}
+		return err
+	}
+	globalKeyCacherOnce.Do(func() {
+		globalKeyCacher = map[string]GlobalCacher{
+			"kid":     {kc: NewMemoryKeyCacher(duration, -1, "kid"), mu: new(sync.RWMutex)},
+			"x5t":     {kc: NewMemoryKeyCacher(duration, -1, "x5t"), mu: new(sync.RWMutex)},
+			"kid_x5t": {kc: NewMemoryKeyCacher(duration, -1, "kid_x5t"), mu: new(sync.RWMutex)},
+		}
+	})
+	return nil
+}
+
+func configGetter(cfg config.ExtraConfig) (time.Duration, error) {
+	e, ok := cfg[ValidatorNamespace].(map[string]interface{})
+	if !ok {
+		return defaultGlobalCacheMaxAge, fmt.Errorf("no config")
+	}
+	duration, ok := e["cache_duration"].(string)
+	if !ok {
+		return defaultGlobalCacheMaxAge, fmt.Errorf("no duration")
+	}
+	return time.ParseDuration(duration)
```

Comment:
Duration should be an uint32 expressing seconds, as we have on the `cache_duration`